src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -182,3 +188,27 @@ def infer_metadata(
             quoted_identifiers = None
 
     return PlanMetadata(attributes=attributes, quoted_identifiers=quoted_identifiers)
+
+
+def cache_metadata_on_select_statement(
+    source_plan: Optional[LogicalPlan], metadata: PlanMetadata
+) -> None:
+    """
+    Cache metadata on a SelectStatement source plan.
+    """
+    from snowflake.snowpark._internal.analyzer.select_statement import SelectStatement
+
+    if (
+        isinstance(source_plan, SelectStatement)
+        and source_plan.analyzer.session.reduce_describe_query_enabled
+    ):
+        source_plan._attributes = metadata.attributes
+        # When source_plan has the same metadata as its child plan,
+        # we should cache it on the child plan too.
+        # This is necessary SelectStatement.select() will need the column states of the child plan
+        # (check the implementation of derive_column_states_from_subquery().
+        if (
+            is_select_statement_child_metadata_same(source_plan)
+            and source_plan.from_._snowflake_plan is not None
+        ):
+            source_plan.from_._snowflake_plan._metadata = metadata
```

Comment:
does this need to be propagated down recursively? `source_plan.from_.from_.snowflake_plan._metadata`?

Reply:
Theoretically yes, but we shouldn't need it because the duplicate query is triggered here https://github.com/snowflakedb/snowpark-python/blob/cf454e9813bcb494c23093f04b4668acc8bf538d/src/snowflake/snowpark/_internal/analyzer/select_statement.py#L719 because source_plan.from_ doesn't have a column states. But if the metadata is cached, then it won't go to next level

Reply:
I guess not since we only come here when projection is None.